wcfsetup/install/files/lib/acp/form/LanguageEditForm.class.php

```diff
@@ -31,79 +29,18 @@ public function readParameters()
     {
         parent::readParameters();
 
-        if (isset($_REQUEST['id'])) {
-            $this->languageID = \intval($_REQUEST['id']);
-        }
-        $this->language = new Language($this->languageID);
-        if (!$this->language->languageID) {
+        $queryParameters = Helper::mapQueryParameters(
+            $_GET,
+            <<<'EOT'
+                array {
+                    id?: positive-int
```

Comment:
I likely wasn't sufficiently clear: `id?` implies that the `id` parameter may be absent. You will still need to check `isset($queryParameters['id'])` or use `$queryParameters['id'] ?? 0` when constructing the DBO.

The reason I suggested making the `id` optional here and manually handling that case is that the Valinor mapping errors are not particularly pretty:

```
{
    "errors": [
        "Value *missing* is not a valid positive integer."
    ]
}
```

@dtdesign Which error handling would you prefer here? Making the `id` required within the array-shape and thus generating the unpretty Valinor error if the `id` is absent – or making the `id` optional, checking it manually and throwing `IllegalLinkException`?

Reply:
The `id` should not be optional, because it is a required part of the route. Making it optional for the sake of avoiding a less helpful error message generated by Valinor is the exact opposite of what I would do.

If you believe that the case of a missing parameter should yield a better error message, then I suggest raising an issue with Valinor to improve this instead.

Reply:
> If you believe that the case of a missing parameter should yield a better error message, then I suggest raising an issue with Valinor to improve this instead.

@dtdesign 

To clarify, the error handling is our own code and I didn't yet check all the capabilities. It primarily exists to have *something*.

https://github.com/WoltLab/WCF/blob/a87fd5826477d6b3579abfb1fd6e01d15ca44be0/wcfsetup/install/files/lib/http/middleware/HandleValinorMappingErrors.class.php#L34-L44

In this case I was concerned about the output, because this is a non-API route, thus the URL is immediately visible in the URL bar and might accidentally be truncated by a user, thus resulting in the JSON output + HTTP 400 instead of the regular `IllegalLinkException` + HTTP 404.

We might be able to generate some well-formatted HTML output within that middleware by switching the output type based on the client's `Accept` header, but the status code cannot be handled generically.

Reply:
I can understand your standpoint, so maybe Valinor is not the right choice there? At the end of the day, marking the parameter as optional here (when it actually is not!) is merely a bogus workaround. Consider this, you’re not marking it as optional because _it is_ optional, but because you want a nicer error handling.

Reply:
@mutec We've discussed this in chat. Please use the following pattern:

```
        try {
            $queryParameters = Helper::mapQueryParameters(
                $_GET,
                <<<'EOT'
                    array {
                        id: positive-int
                    }
                    EOT
            );
            $this->formObject = new Language($queryParameters['id']);

            if (!$this->formObject->getObjectID()) {
                throw new IllegalLinkException();
            }
        } catch (MappingError) {
            throw new IllegalLinkException();
        }
```

It's a little on the verbose side due to the additional try-catch, but it avoids all the pitfalls and issues with parameter validation. If the try-catch is missing it fails “safely”: The error message won't be pretty, but the behavior will be correct.

Reply:
Seeing that code makes me thinking about `DatabaseObject::construct()` throwing an exception in case the id is not `null` and no object is found. :D

Reply:
In hindsight it would have been much better to implement a static method for DBOs that returns the constructed object or null if it was not found. The whole magic behind `$data` is just a mess and if we would design these today, we would simply use a `readonly class` that represents all fields as actual properties.

But this is a story for another day ;)